src/Rules/Functions/ParameterCastableToStringFunctionRule.php

```diff
@@ -0,0 +1,107 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Functions;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Rules\RuleLevelHelper;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\Type;
+use PHPStan\Type\VerbosityLevel;
+use function count;
+use function in_array;
+use function sprintf;
+
+/**
+ * @implements Rule<Node\Expr\FuncCall>
+ */
+class ParameterCastableToStringFunctionRule implements Rule
+{
+
+	public function __construct(
+		private ReflectionProvider $reflectionProvider,
+		private RuleLevelHelper $ruleLevelHelper,
+	)
+	{
+	}
+
+	public function getNodeType(): string
+	{
+		return FuncCall::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (!($node->name instanceof Node\Name)) {
+			return [];
+		}
+
+		$functionName = $this->reflectionProvider->resolveFunctionName($node->name, $scope);
+		$args = $node->getArgs();
+		$errorMessage = 'Parameter #%d of function %s expects an array of values castable to string, %s given.';
+		if (in_array($functionName, ['implode', 'join'], true)) {
+			$errorMessage = 'Parameter #%d $array of function %s expects array<string>, %s given.';
+			if (count($args) === 1) {
+				$argsToCheck = [0 => $args[0]];
+			} elseif (count($args) === 2) {
+				$argsToCheck = [1 => $args[1]];
```

Comment:
I know you probably just copied this from ImplodeFunctionRule, but the assumption is wrong because of named arguments. Someone might call `implode(array: $a, separator: $s)`. and this code would break.

Please write a test for this scenario. The scenario can be fixed by processing `$args` through `ArgumentsNormalizer`.

Reply:
Good catch - I didn't think of that, since I just modifed the ImplodeFunctionRule. There was also a similar issue when no arguments were passed etc.

----

However, it lead me down another rabbit hole: when I added support for named arguments, I had to do something with the `Parameter #1 of function ...`, since the number might now be incorrect. So I attempted to make it consistent with how `FunctionCallParametersCheck` names the parameters. But it feels like individual rules shouldn't deal with that at all, since then it will "never" be consistent. If this is already handled somehow, maybe you can point me in the right direction.

Reply:
Yeah, other rules don't deal with this at all now. We could refactor FunctionCallParametersCheck to be more flexible and to use its internals in other rules, but I don't think it's necessary.

Another thing this would allow us is to deal with `$arg–>unpack` properly same way FunctionCallParametersCheck does - it actually looks into `$type->getConstantArrays()` and expands them: https://github.com/phpstan/phpstan-src/blob/38f390636cb601b02071d3956c80d64f2f4b362b/src/Rules/FunctionCallParametersCheck.php#L114-L162